examples/platform/esp32/route_hook/esp_route_hook.c

```diff
@@ -0,0 +1,199 @@
+                memcpy(&prefix.addr, rio_data, prefix_len_bytes);
+                route.netif            = netif;
+                route.gateway          = *src_addr;
+                route.prefix_length    = rio_header->prefix_length;
+                route.prefix           = prefix;
+                route.preference       = preference;
+                route.lifetime_seconds = lwip_ntohl(rio_header->route_lifetime);
+                ESP_LOGI(TAG, "prefix %s lifetime %u\n", ip6addr_ntoa(&prefix), route.lifetime_seconds);
+                if (esp_route_table_add_route_entry(&route) == NULL)
+                {
+                    ESP_LOGI(TAG, "Failed to add route table entry\n");
+                }
+            }
+        }
+        icmp_payload += opt_len;
+        payload_len -= opt_len;
+    }
+}
+
+static uint8_t icmp6_raw_recv_handler(void * arg, struct raw_pcb * pcb, struct pbuf * p, const ip_addr_t * addr)
+{
+    uint8_t * icmp_payload = NULL;
+    uint16_t icmp_payload_len;
+    struct ip6_hdr * ip6_header = (struct ip6_hdr *) p->payload;
+    struct icmp6_hdr * icmp6_header;
+    ip6_addr_t src;
+    ip6_addr_t dest;
+    esp_route_hook * hook = (esp_route_hook *) arg;
+
+    memcpy(src.addr, ip6_header->src.addr, sizeof(src.addr));
+    memcpy(dest.addr, ip6_header->dest.addr, sizeof(dest.addr));
+#if LWIP_IPV6_SCOPES
+    src.zone = 0;
+#endif
+
+    if (p->tot_len != p->len)
+    {
+        ESP_LOGW(TAG, "Ignore segmented ICMP packet");
+        return 0;
+    }
+    if (p->tot_len <= sizeof(struct ip6_hdr) + sizeof(struct icmp6_hdr))
+    {
+        ESP_LOGW(TAG, "Ignore invalid ICMP packet");
+        return 0;
+    }
+    if (!ip6_addr_islinklocal(&dest) && !ip6_addr_isallnodes_linklocal(&dest) && !ip6_addr_isallrouters_linklocal(&dest))
+    {
+        return 0;
+    }
+
+    icmp_payload_len = p->tot_len - sizeof(struct ip6_hdr);
+    icmp_payload     = p->payload + sizeof(struct ip6_hdr);
+
+    icmp6_header = (struct icmp6_hdr *) icmp_payload;
+    if (icmp6_header->type == ICMP6_TYPE_RA)
+    {
+        ra_recv_handler(hook->netif, icmp_payload, icmp_payload_len, &src);
+    }
+    return 0;
+}
+
+esp_err_t esp_route_hook_init(esp_netif_t * netif)
+{
+    struct netif * lwip_netif;
+    ip_addr_t router_group = IPADDR6_INIT_HOST(0xFF020000, 0, 0, 0x02);
+    esp_route_hook * hook  = NULL;
+    esp_err_t ret          = ESP_OK;
+
+    ESP_RETURN_ON_FALSE(netif != NULL, ESP_ERR_INVALID_ARG, TAG, "Invalid network interface");
+    lwip_netif = netif_get_by_index(esp_netif_get_netif_impl_index(netif));
+    ESP_RETURN_ON_FALSE(lwip_netif != NULL, ESP_ERR_INVALID_ARG, TAG, "Invalid network interface");
+
+    for (esp_route_hook * iter = s_hooks; iter != NULL; iter++)
+    {
+        if (iter->netif == lwip_netif)
+        {
+            ESP_LOGI(TAG, "Hook already installed on netif, skip...");
+            return ESP_OK;
+        }
+    }
+
+    hook = (esp_route_hook *) malloc(sizeof(esp_route_hook));
```

Comment:
If all the members of hook are not set by you, then better use `calloc` or explicitly memset hook?

Reply:
All the hook members are set in the following lines so it's fine to use `malloc` here.